core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java

```diff
@@ -0,0 +1,267 @@
+        }
+
+        public Response getResponse() {
+            return response;
+        }
+
+        public Translog.Location getLocation() {
+            return location;
+        }
+    }
+
+    /**
+     * Result of taking the action on the primary.
+     */
+    private class WritePrimaryResult extends PrimaryResult implements RespondingWriteResult<Response> {
+        private final ReplicatedWriteRequest<?> request;
+        volatile boolean finishedWrite;
+        volatile ActionListener<Response> listener = null;
+
+        public WritePrimaryResult(Request request, Response finalResponse,
+                                  @Nullable Translog.Location location,
+                                  IndexShard indexShard) {
+            super(request, finalResponse);
+            this.request = request;
+            /*
+             * We call this before replication because this might wait for a refresh and that can take a while. This way we wait for the
+             * refresh in parallel on the primary and on the replica.
+             */
+            finishWrite(indexShard, request, location);
+        }
+
+        @Override
+        public synchronized void respond(ActionListener<Response> listener) {
+            this.listener = listener;
+            respondIfPossible();
+        }
+
+        /**
+         * Respond if the refresh has occurred and the listener is ready. Always called while synchronized on {@code this}.
+         */
+        protected void respondIfPossible() {
+            if (finishedWrite && listener != null) {
+                super.respond(listener);
+            }
+        }
+
+        @Override
+        public void forcedRefresh() {
+            finalResponse.setForcedRefresh(true);
+        }
+
+        @Override
+        public void waitForRefreshForcedRefresh() {
+            logger.warn("block_until_refresh request ran out of slots and forced a refresh: [{}]", request);
+        }
+
+        @Override
+        public synchronized void finished() {
+            finishedWrite = true;
+            respondIfPossible();
+        }
+    }
+
+    /**
+     * Result of taking the action on the replica.
+     */
+    private class WriteReplicaResult extends ReplicaResult implements RespondingWriteResult<TransportResponse.Empty> {
+        private final IndexShard indexShard;
+        private final ReplicatedWriteRequest<?> request;
+        private final Translog.Location location;
+        private ActionListener<TransportResponse.Empty> listener;
+
+        public WriteReplicaResult(IndexShard indexShard, ReplicatedWriteRequest<?> request, Translog.Location location) {
+            this.indexShard = indexShard;
+            this.request = request;
+            this.location = location;
+        }
+
+        @Override
+        public void respond(ActionListener<TransportResponse.Empty> listener) {
+            this.listener = listener;
+            finishWrite(indexShard, request, location);
```

Comment:
I think it's confusing that the WriteReplicaResult flow is different than WritePrimaryResul. i.e., `finishWrite` is called in the constructor for one and in the respond method for the other. We should try to make them the same as far as possible.